network/network-address/src/lib.rs

```diff
@@ -0,0 +1,619 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use lcs;
+use libra_crypto::{
+    traits::{CryptoMaterialError, ValidKeyStringExt},
+    x25519,
+};
+#[cfg(any(test, feature = "fuzzing"))]
+use proptest::{collection::vec, prelude::*};
+#[cfg(any(test, feature = "fuzzing"))]
+use proptest_derive::Arbitrary;
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+use std::{
+    convert::{Into, TryFrom},
+    fmt,
+    net::{self, Ipv4Addr, Ipv6Addr},
+    num,
+    str::FromStr,
+    string::ToString,
+};
+use thiserror::Error;
+
+const MAX_DNS_NAME_SIZE: usize = 255;
+
+/// A `RawNetworkAddress` is the serialized, unverified, on-chain representation
+/// of a [`NetworkAddress`]. Specifically, a `RawNetworkAddress` is usually an
+/// [`lcs`]-serialized `NetworkAddress`.
+///
+/// This representation is useful because:
+///
+/// 1. Move does't understand (and doesn't really need to understand) what a
+///    `NetworkAddress` is, so we can just store an opaque `vector<u8>` on-chain,
+///    which we represent as `RawNetworkAddress` in Rust.
+/// 2. We want to deserialize a `Vec<NetworkAddress>` but ignore ones that don't
+///    properly deserialize. For example, a validator might advertise several
+///    `NetworkAddress`. If that validator is running a newer version of the node
+///    software, and old versions can't understand one of the `NetworkAddress`,
+///    they would be able to ignore the one that doesn't deserialize for them.
+///    We can easily do this by storing a `vector<vector<u8>>` on-chain, which
+///    we deserialize as `Vec<RawNetworkAddress>` in Rust. Then we deserialize
+///    each `RawNetworkAddress` into a `NetworkAddress` individually.
+///
+/// Note: deserializing a `RawNetworkAddress` does no validation, other than
+/// deserializing the underlying `Vec<u8>`.
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub struct RawNetworkAddress(#[serde(with = "serde_bytes")] Vec<u8>);
+
+/// Libra `NetworkAddress` is a compact, efficient, self-describing and
+/// future-proof network address represented as a stack of protocols. Essentially
+/// libp2p's [multiaddr](https://multiformats.io/multiaddr/) but using [`lcs`] to
+/// describe the binary format.
+///
+/// Most validators will advertise a network address like:
+///
+/// `/dns/example.com/tcp/6180/ln-noise-ik/<x25519-pubkey>/ln-handshake/1`
+///
+/// Unpacking, the above effectively means:
+///
+/// 1. Resolve the DNS name "example.com" to an ip address, `addr`.
+/// 2. Open a TCP connection to `(addr, 6180)`.
+/// 3. Perform a Noise IK handshake and assume the peer's static pubkey is
+///    `<x25519-pubkey>`. After this step, we will have a secure, authenticated
+///    connection with the peer.
```

Comment:
A few minor nit (maybe too late for that):

* I don't really like noise being used in defining the protocol because noise is not a protocol by itself. For example, we use noise IK without payloads and with a timestamp as the prologue.
* we have two handshakes now :) 

I think it'd be nicer to have something like this:

`/dns/example.com/tcp/6180/libranet_v1/<x25519-pubkey>`